nucliadb/src/nucliadb/common/cluster/rebalance.py

```diff
@@ -153,27 +193,189 @@ async def move_set_of_kb_resources(
+        async with (
+            locking.distributed_lock(locking.NEW_SHARD_LOCK.format(kbid=kbid)),
+            datamanagers.with_rw_transaction() as txn,
+        ):
+            sm = get_shard_manager()
+            for _ in range(shards_to_add):
+                await sm.create_shard_by_kbid(txn, kbid)
+            await txn.commit()
+
+        # Recalculate after having created shards, the active shard is a different one
+        shards = await get_rebalance_shards(kbid)
+
+    # Now, move resoures to other shards as long as we are still over the max
+    with_room = [
+        s for s in shards if s.id != shard.id and s.paragraphs < settings.max_shard_paragraphs * 0.9
+    ]
+    shard_paragraphs = shard.paragraphs
+    i = 0
+    while (shard_paragraphs > settings.max_shard_paragraphs) and i < 500:
+        try:
+            # Get the biggest shard with some space left
+            target_shard = with_room[-1]
+        except IndexError:
+            logger.warning("No target shard found for splitting", extra={"kbid": kbid})
             break
 
-        await move_set_of_kb_resources(context, kbid, largest_shard, smallest_shard)
+        await move_set_of_kb_resources(
+            context, kbid, from_shard_id=shard.id, to_shard_id=target_shard.id, count=20
+        )
+
+        # Recalculate shard paragraphs
+        shards = await get_rebalance_shards(kbid)
+        shard_paragraphs = next(s.paragraphs for s in shards if s.id == shard.id)
+        with_room = [
+            s for s in shards if s.id != shard.id and s.paragraphs < settings.max_shard_paragraphs * 0.9
+        ]
+        i += 1
 
-        rebalanced_shards.add(largest_shard)
 
-        shard_paragraphs = await get_shards_paragraphs(kbid)
+async def merge_shard(
+    context: ApplicationContext, kbid: str, shard: RebalanceShard, shards: list[RebalanceShard]
+):
+    logger.info(
+        "Merging shard",
+        extra={
+            "kbid": kbid,
+            "shard": shard.to_dict(),
+        },
+    )
+    empty_shard = False
+
+    i = 0
+    while i < 1000:
+        resources_count = await get_shard_resources_count(shard.nidx_id)
+        if resources_count == 0:
+            logger.info(
+                "Shard is now empty",
+                extra={
+                    "kbid": kbid,
+                    "shard": shard.to_dict(),
+                },
+            )
+            empty_shard = True
+            break
+
+        logger.info(
+            "Shard not yet empty",
+            extra={
+                "kbid": kbid,
+                "shard": shard.to_dict(),
+                "remaining": resources_count,
+            },
+        )
+
+        # Take the biggest shard that has some capacity (90% of the max, to prevent filling them completely)
+        try:
+            with_room = [
+                s
+                for s in shards
+                if s.id != shard.id and s.paragraphs < settings.max_shard_paragraphs * 0.9
+            ]
```

Comment:
Shouldn't you check if the shards are <= 90% of the max as the comment says?

Reply:
that is what I am doing no? oh you mean I should add the <= instead of just <?

Reply:
Oh weird, I may have looked at an outdated version